### PR TITLE
Feature/pre commit hooks black isort

### PR DESCRIPTION
**Pull request by ZhukovGreen**
https://issues.redhat.com/browse/OAMG-4765

TODO:
Acceptance criteria:
- [x] pre-commit config present in the project tree, and contains two hooks : black and isort
- [x] configurations of black and isort placed in pyproject.toml
- [x] Add a workflow of running pre-commit (pre-commit run --all-files) into the githab actions for cetnos8 image only
- [x] Document the work with the pre-commit hooks
- [x] Test with basic convertability tests

Notes for the reviewer:
Review by commits. 
https://github.com/oamg/convert2rhel/pull/264/commits/5a2aff98ffde5c6048bbdeaf295ba22af9bf4e49 introduces configs
https://github.com/oamg/convert2rhel/pull/264/commits/e52858e87906bd1a20f295e8a55bb88bf40d01a3 - introduces changes by the pre-commit

### CODE REVIEW

**Comment by ZhukovGreen** [comment]
@oamg/convert2rhel please review

**Comment by ZhukovGreen** [comment]
ref #153

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,97 @@
+[tool.black]
+line-length=119
```
Comment (ZhukovGreen) [inline pyproject.toml:2]: I would vote for 79 line length. 119 - is not readable and too long for me.
@oamg/convert2rhel wdyt?

**Comment by ZhukovGreen** [comment]
Tested on CentOS 6,7,8 and OL 6, 7,8

File: pyproject.toml (hunk 1 above)
Comment (fellipeh) [reply pyproject.toml:2]: @ZhukovGreen  I just think 79 is too small... in some cases (eg: conditionals) will be weird to read.. I vote between this two: 88 line length as `black` default.

**Review by fellipeh — COMMENTED** [review]
@ZhukovGreen  we'll some command to run isort into our files before the commit ?

**Comment by ZhukovGreen** [comment]
Yes, similar to black

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+minimum_pre_commit_version: "2.9.0"
```
Comment (bocekm) [inline .pre-commit-config.yaml:1]: Why is this needed?

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+repos:
+  - repo: "https://github.com/psf/black"
+    rev: "21.4b2"
```
Comment (bocekm) [inline .pre-commit-config.yaml:4]: Can't we just omit the `rev` to always get the latest version?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -6,8 +6,11 @@
 
 ## Developing locally
 
-The commands below installs a python3 virtual environment
-with all the necessary dependencies installed:
+The commands below creates a python3 virtual environment,
+with all the necessary dependencies installed,
+builds needed images,
+creates dotenv file, and
```
Comment (bocekm) [inline CONTRIBUTING.md:12]:
```suggestion
The commands below create a python3 virtual environment with all the necessary dependencies installed,
build needed images,
create .env file, and
```

File: pyproject.toml (hunk 1 above)
Comment (bocekm) [reply pyproject.toml:2]: It was 80 when I started. Then we decided for 120 about a year ago. So I vote for not getting back to the discussion and stick to the length we have now. I can find the PR where we discussed it if you want.

```suggestion
line-length=120
```

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+[tool.coverage.run]
+branch = false
+source = ["convert2rhel"]
+omit = ["convert2rhel/unit_tests"]
+
+[tool.coverage.paths]
+source = ["convert2rhel/"]
+
+[tool.coverage.report]
+fail_under = 87
```
Comment (bocekm) [inline pyproject.toml:28]: Last week we moved it up to 80. Are you changing the coverage check so that it reports 87% now?

File: pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    "^get.*$",
+    "^factory$",
+    "^_.+$",
+]
+verbose = 2
+quiet = false
+whitelist-regex = []
+color = true
+
+[tool.poetry]
```
Comment (bocekm) [inline pyproject.toml:64]: What is poetry? How can I verify all its setup is correct?

File: pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,97 @@
+[tool.black]
+line-length=119
+target-version=["py38"]
```
Comment (bocekm) [inline pyproject.toml:3]: Are you confident that the black formatting for Python 3.8 won't use formatting not compatible with Python 2.6?

**Review by bocekm — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (ZhukovGreen) [reply pyproject.toml:2]: Ok, no need to block on this. Applied

File: .pre-commit-config.yaml (hunk 2 above)
Comment (ZhukovGreen) [reply .pre-commit-config.yaml:1]: This is similar to the need of pinning versions in requirements.txt. If we allow having the latest version, then we can't guarantee reproducibility. In case of pre-commit it is even better - you just run:
pre-commit auto-update and these versions are updated automatically

File: .pre-commit-config.yaml (hunk 3 above)
Comment (ZhukovGreen) [reply .pre-commit-config.yaml:4]: the same as before

File: CONTRIBUTING.md (hunk 4 above)
Comment (ZhukovGreen) [reply CONTRIBUTING.md:12]: applied

File: pyproject.toml (hunk 6 above)
Comment (ZhukovGreen) [reply pyproject.toml]: that was added accidentally - now no poetry here

File: pyproject.toml (hunk 7 above)
Comment (ZhukovGreen) [reply pyproject.toml:3]: Tested with py26 unit tests and centos6 vagrant box

File: pyproject.toml (hunk 5 above)
Comment (ZhukovGreen) [reply pyproject.toml]: you're right, made a little mistake in migrating the config. Thx

**Review by bocekm — APPROVED** [review]

**Review by fellipeh — APPROVED** [review]

**Review by bocekm — APPROVED** [review]

File: pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,30 @@
+[tool.black]
+line-length=120
+target-version=["py38"]
+
+[tool.isort]
```
Comment (SpyTec) [inline pyproject.toml:5]: @ZhukovGreen I noticed this in the doc

https://pycqa.github.io/isort/docs/configuration/black_compatibility/

https://pycqa.github.io/isort/docs/configuration/profiles/

From that we're missing this, should we add that or do the black/isort not collide with each other with current config?
```
force_grid_wrap: 0
ensure_newline_before_comments: true
```

Comment (ZhukovGreen) [reply pyproject.toml:5]: They are not collide in this particular setup. Time will show, however in my projects I never seen the need to solve the conflict. Time will show.

Comment (bocekm) [reply pyproject.toml:5]: I like Eric's suggestion. It would make the config a bit more concise as well:
```
atomic=true
line_length = 120
lines_after_imports=2
lines_between_types=1
use_parentheses = true
balanced_wrapping = true
include_trailing_comma = true
multi_line_output = 3
```
down to:
```
profile = "black"
atomic = true
line_length = 120
lines_after_imports = 2
lines_between_types = 1
balanced_wrapping = true
```

Comment (ZhukovGreen) [reply pyproject.toml:5]: OK, should be addressed by #279
